### PR TITLE
Upgrade TypeScript to v4

### PR DESCRIPTION
**Pull request by sirreal**
## Description
Upgrade TypeScript dependency to version 4.

Read the announcement [here](https://devblogs.microsoft.com/typescript/announcing-typescript-4-0/).

Applies a new feature [`noEmit` with `incremental`](https://devblogs.microsoft.com/typescript/announcing-typescript-4-0/#noemit-and-incremental) to the `bin/` directory typings. These typings were discarded before, this change prevents them from being produced at all.

## How has this been tested?

Types continue to be generated as expected. The following completes without issues:

```sh
npm run clean:package-types
npm run build:package-types
npm run build:package-types
```

Types are generated as expected, e.g. `packages/warning/build-types/index.d.ts`:

```ts
/**
 * Shows a warning with `message` if environment is not `production`.
 *
 * @param {string} message Message to show in the warning.
 *
 * @example
 * ```js
 * import warning from '@wordpress/warning';
 *
 * function MyComponent( props ) {
 *   if ( ! props.title ) {
 *     warning( '`props.title` was not passed' );
 *   }
 *   ...
 * }
 * ```
 */
export default function warning(message: string): void;
//# sourceMappingURL=index.d.ts.map
```

## Types of changes

Internal.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: bin/tsconfig.json [hunk 1]
```diff
@@ -6,11 +6,13 @@
 		"target": "ES6",
 		"lib": [ "ES6", "ES2020.string" ],
 		"rootDir": ".",
-		"declarationMap": false,
 
-		// We're not interested in the output, but we must generate
-		// something as part of a composite project. Use the
-		// ignored `.cache` file to hide tsbuildinfo and d.ts files.
+		// These script are unpublished and packages do not depend on them.
```
Comment (ZebulanStanphill) [inline bin/tsconfig.json:10]:
```suggestion
		// These scripts aren't published and packages don't depend on them.
```

**Comment by sirreal** [comment]
Rebased and upgraded to latest 4.0.3.

File: bin/tsconfig.json (hunk 1 above)
Comment (ockham) [reply bin/tsconfig.json:10]: @sirreal Did you mean to commit @ZebulanStanphill's suggestion? Cause it seems to be _uncommitted_ :wink:

**Review by ockham — APPROVED** [review]
Looking good and testing well! [One note](https://github.com/WordPress/gutenberg/pull/24892/files#r499669147), otherwise good to merge!

File: bin/tsconfig.json (hunk 1 above)
Comment (sirreal) [reply bin/tsconfig.json:10]: Yes, I think I lost it in a rebase 🤦 

Thanks!

**Pull request merged**